### DISCUSSION TITLE
Asyncua Errors - BadtypeMismatch

### DISCUSSION THREAD

**Discussion by andycg89** [FreeOpcUa/opcua-asyncio#1546] [General]
When I run this Script I keep getting the mismatch error. Any insights? New to this. Thanks in advance!

`import logging
import asyncio
import pandas as pd
import nest_asyncio

from asyncua import ua, Server
from asyncua.common.methods import uamethod

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger('asyncua')

@uamethod
def func(parent, value):
    return value * 2

async def main():
    # setup our server
    server = Server()
    await server.init()
    server.set_endpoint('opc.tcp://127.0.0.1:4843/opcua/')
    server.set_server_name("DevNet OPC-UA Test Server")

    # setup our own namespace, not really necessary but should as spec
    uri = 'http://devnetiot.com/opcua/'
    idx = await server.register_namespace(uri)

    # populating our address space
    # server.nodes, contains links to very common nodes like objects and root
    obj_vplc = await server.nodes.objects.add_object(idx, 'vPLC1')
    var_temperature = await obj_vplc.add_variable(idx, 'temperature', 0)
    var_pressure = await obj_vplc.add_variable(idx, 'pressure', 0)
    var_pumpsetting = await obj_vplc.add_variable(idx, 'pumpsetting', 0)

    # Read Sensor Data from Kaggle
    df = pd.read_csv("sensor.csv")
    # Only use sensor data from 03 and 01 (preference)
    sensor_data = pd.concat([df["sensor_03"], df["sensor_01"]], axis=1)

    _logger.info('Starting server!')
    async with server:
        # run forever and iterate over the dataframe
        while True:
            for row in sensor_data.itertuples():
                # if below the mean use different setting - just for testing
                if row[1] < df["sensor_03"].mean():
                    setting = "standard"
                else:
                    setting = "speed"

                # Writing Variables
                await var_temperature.write_value(float(row[1]))
                await var_pressure.write_value(float(row[2]))
                await var_pumpsetting.write_value(str(setting))
                await asyncio.sleep(1)

if __name__ == '__main__':
    #python 3.6 or lower
   # loop = asyncio.get_event_loop() #- deprecated
    #loop = get_event_loop()
   # loop.run_until_complete(main())
    #python 3.7 onwards (comment lines above)
    #asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy()) #Addition from Web
   # nest_asyncio.apply()
    asyncio.run(main())`

OUTPUT FROM CODE:

INFO:asyncua.server.internal_server:No user manager specified. Using default permissive manager instead.
INFO:asyncua.server.internal_session:Created internal session Internal
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=11715, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15958, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15959, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15960, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15961, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15962, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15963, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=15964, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=16134, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=16135, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua.server.address_space:add_node: while adding node NumericNodeId(Identifier=16136, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>), requested parent node NumericNodeId(Identifier=15957, NamespaceIndex=0, NodeIdType=<NodeIdType.Numeric: 2>) does not exists
INFO:asyncua:Starting server!
WARNING:asyncua.server.server:Endpoints other than open requested but private key and certificate are not set.
INFO:asyncua.server.internal_server:starting internal server
INFO:asyncua.server.binary_server_asyncio:Listening on 127.0.0.1:4843
WARNING:asyncua.server.address_space:Write refused: Variant: Variant(Value=46.31076, VariantType=<VariantType.Double: 11>, Dimensions=None, is_array=False) with type VariantType.Double does not have expected type: VariantType.Int64
INFO:asyncua.server.binary_server_asyncio:Closing asyncio socket server
INFO:asyncua.server.internal_server:stopping internal server
INFO:asyncua.server.internal_session:close session Internal
INFO:asyncua.server.subscription_service:delete subscriptions: []
Traceback (most recent call last):
  File "C:\Users\andyc\.vscode\CNC\.venv\opc-ua-sensor-simulator\opc-ua-server-copy.py", line 71, in <module>
    asyncio.run(main())
  File "C:\Users\andyc\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\andyc\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\andyc\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 684, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\andyc\.vscode\CNC\.venv\opc-ua-sensor-simulator\opc-ua-server-copy.py", line 58, in main
    await var_temperature.write_value(float(row[1]))
  File "C:\Users\andyc\.vscode\CNC\.venv\Lib\site-packages\asyncua\common\node.py", line 269, in write_value
    await self.write_attribute(ua.AttributeIds.Value, dv)
  File "C:\Users\andyc\.vscode\CNC\.venv\Lib\site-packages\asyncua\common\node.py", line 323, in write_attribute
    result[0].check()
  File "C:\Users\andyc\.vscode\CNC\.venv\Lib\site-packages\asyncua\ua\uatypes.py", line 375, in check
    raise UaStatusCodeError(self.value)
asyncua.ua.uaerrors._auto.BadTypeMismatch: The value supplied for the attribute is not of the same type as the attribute"s value.(BadTypeMismatch)

**Comment by schroeder-** [comment]
You have to match the correct datatype. To do this we have types:

- Ints: ua.Int32, ua.Int64, ua.UInt32, ...
- Float: ua.Float, ua.Double

A example how to use:

`
node = server.add_variable(..., ua.Double(0.0))
my_num = float("1.0")
node.write_value(ua.Double(my_num))`
